File: extensions/sdk/python/picteus_extension_sdk/intents.py
import json
from dataclasses import dataclass, asdict, field
from enum import StrEnum
from typing import Dict, Any, Literal, Optional, Union, List

# noinspection PyPackageRequirements

Json = Dict[str, Any]


# In order to benefit from serializable intent structures, taken from https://stackoverflow.com/questions/51286748/make-the-python-json-encoder-support-pythons-new-dataclasses
@dataclass
class SuperDataClass:

    @property
    def __dict__(self):
        """
        get a Python dictionary
        """
        # noinspection PyTypeChecker
        return asdict(self)

    @property
    def json(self):
        """
        get the JSON formated string
        """
        return json.dumps(self.__dict__)


@dataclass(kw_only=True)
class IntentIdentity(SuperDataClass):
    id: str


@dataclass(kw_only=True)
class IntentContext(SuperDataClass):
    imageIds: Optional[List[str]] = field(default=None)


@dataclass
class BasisIntent(SuperDataClass):
    identity: Optional[IntentIdentity] = field(default=None)


@dataclass
class WithContextIntent(BasisIntent):
    context: Optional[IntentContext] = field(default=None)


@dataclass
class IntentResourceUrl(SuperDataClass):
    url: str


@dataclass
class IntentResourceContent(SuperDataClass):
    content: bytearray


IntentResource = Union[IntentResourceUrl, IntentResourceContent]


@dataclass(kw_only=True)
class IntentDialogContent(SuperDataClass):
    title: str
    description: str
    details: Optional[str] = field(default=None)


@dataclass(kw_only=True)
class IntentDialogIconContent(IntentDialogContent):
    icon: Optional[IntentResource] = field(default=None)


@dataclass
class IntentDialogIconSizeContent(IntentDialogIconContent):
    size: Optional[Literal["auto", "xs", "s", "m", "l", "xl"]] = field(default=None)


@dataclass
class IntentFormContent(SuperDataClass):
    parameters: Json
    dialogContent: Optional[IntentDialogIconSizeContent] = field(default=None)


@dataclass(kw_only=True)
class FormIntent(WithContextIntent):
    form: IntentFormContent


class IntentUiAnchor(StrEnum):
    MODAL = "modal"
    SIDEBAR = "sidebar"
    WINDOW = "window"
    IMAGE_DETAILS = "imageDetail"


@dataclass(kw_only=True)
class IntentUISidebarIntegration(SuperDataClass):
    anchor: IntentUiAnchor = field(default=IntentUiAnchor.SIDEBAR, init=False)
    isExternal: bool


@dataclass(kw_only=True)
class IntentUIWindowIntegration(SuperDataClass):
    anchor: IntentUiAnchor = field(default=IntentUiAnchor.WINDOW, init=False)


@dataclass(kw_only=True)
class IntentUIModalIntegration(SuperDataClass):
    anchor: IntentUiAnchor = field(default=IntentUiAnchor.MODAL, init=False)


IntentUIIntegration = Union[
    IntentUISidebarIntegration, IntentUIWindowIntegration, IntentUIModalIntegration]


@dataclass
class IntentFrameUrlContent(SuperDataClass):
    url: str


@dataclass
class IntentFrameHtmlContent(SuperDataClass):
    html: str


IntentFrameContent = Union[IntentFrameUrlContent, IntentFrameHtmlContent]


@dataclass
class IntentUi(SuperDataClass):
    id: str
    integration: IntentUIIntegration
    frameContent: IntentFrameContent
    dialogContent: Optional[IntentDialogIconContent] = field(default=None)


@dataclass(kw_only=True)
class UiIntent(WithContextIntent):
    ui: IntentUi


class IntentDialogType(StrEnum):
    ERROR = "error"
    INFO = "info"
    QUESTION = "question"


@dataclass
class IntentFrame(SuperDataClass):
    content: IntentFrameContent
    height: int


@dataclass
class IntentDialogButtons(SuperDataClass):
    yes: str
    no: Optional[str] = field(default=None)


@dataclass(kw_only=True)
class IntentDialog(IntentDialogIconSizeContent):
    type: IntentDialogType
    frame: Optional[IntentFrame] = field(default=None)
    buttons: IntentDialogButtons


@dataclass(kw_only=True)
class DialogIntent(WithContextIntent):
    dialog: IntentDialog


@dataclass
class IntentImage(SuperDataClass):
    imageId: str
    dialogContent: Optional[IntentDialogContent] = field(default=None)


@dataclass
class IntentImages(SuperDataClass):
    images: List[IntentImage]
    dialogContent: Optional[IntentDialogIconContent] = field(default=None)


@dataclass(kw_only=True)
class ImagesIntent(WithContextIntent):
    images: IntentImages


class IntentShowType(StrEnum):
    SIDEBAR = "sidebar"
    EXTENSION_SETTINGS = "extensionSettings"
    IMAGE = "image"
    REPOSITORY = "repository"


@dataclass
class IntentShow(SuperDataClass):
    type: IntentShowType
    id: str


@dataclass(kw_only=True)
class ShowIntent(BasisIntent):
    show: IntentShow


class IntentToastType(StrEnum):
    INFO = "info"
    CANCEL = "cancel"
    ERROR = "error"


@dataclass(kw_only=True)
class IntentToast(SuperDataClass):
    type: IntentToastType
    title: Optional[str] = field(default=None)
    subtitle: str


@dataclass(kw_only=True)
class ToastIntent(WithContextIntent):
    toast: IntentToast


@dataclass(kw_only=True)
class IntentNotification(SuperDataClass):
    title: str
    subtitle: str
    body: str
    silent: bool
    icon: Optional[bytearray] = field(default=None)
    isNative: bool


@dataclass(kw_only=True)
class NotificationIntent(WithContextIntent):
    notification: IntentNotification


@dataclass
class IntentProcessCommand(SuperDataClass):
    extensionId: str
    commandId: str


@dataclass(kw_only=True)
class ProcessCommandIntent(WithContextIntent):
    processCommand: IntentProcessCommand


@dataclass
class IntentAction(SuperDataClass):
    intent: Union[ShowIntent, UiIntent, ProcessCommandIntent]
    dialogContent: IntentDialogIconSizeContent
    label: Optional[str] = field(default=None)


@dataclass(kw_only=True)
class ActionIntent(WithContextIntent):
    action: IntentAction


FrontIntent = Union[
    FormIntent, UiIntent, DialogIntent, ImagesIntent, ShowIntent, ToastIntent, NotificationIntent, ActionIntent]


@dataclass(kw_only=True)
class IntentServeBundle(SuperDataClass):
    content: bytearray
    settings: Optional[Json] = field(default=None)


@dataclass(kw_only=True)
class ServeBundleIntent(BasisIntent):
    serveBundle: IntentServeBundle


@dataclass(kw_only=True)
class IntentReadFile(SuperDataClass):
    extensions: Optional[List[str]] = field(default=None)
    message: str


@dataclass(kw_only=True)
class ReadFileIntent(WithContextIntent):
    readFile: IntentReadFile


@dataclass
class IntentWriteFile(SuperDataClass):
    content: bytearray
    name: str
    extension: str
    message: str


@dataclass(kw_only=True)
class WriteFileIntent(WithContextIntent):
    writeFile: IntentWriteFile


BackIntent = Union[
    ServeBundleIntent, ReadFileIntent, WriteFileIntent]

Intent = Union[FrontIntent, BackIntent]

File: extensions/sdk/typescript/src/intents.ts
type Json = Record<string, any>;

export interface IntentIdentity
{
  readonly id: string;
}

export interface IntentContext
{
  readonly imageIds?: string[];
}

export interface BasisIntent
{
  readonly identity?: IntentIdentity;
}

export interface WithContextIntent extends BasisIntent
{
  readonly context?: IntentContext;
}

export interface IntentResourceUrl
{
  readonly url: string;
}

export interface IntentResourceContent
{
  readonly content: Buffer;
}

export type IntentResource = IntentResourceUrl | IntentResourceContent;

export interface IntentDialogContent
{
  readonly title: string;
  readonly description: string;
  readonly details?: string;
}

export interface IntentDialogIconContent extends IntentDialogContent
{
  readonly icon?: IntentResource;
}

export interface IntentDialogIconSizeContent extends IntentDialogIconContent
{
  readonly size?: "auto" | "xs" | "s" | "m" | "l" | "xl";
}

export interface IntentFormContent
{
  readonly parameters: Json;
  readonly dialogContent?: IntentDialogIconSizeContent;
}

export interface FormIntent extends WithContextIntent
{
  readonly form: IntentFormContent;
}

export enum IntentUiAnchor
{
  Modal = "modal",
  Sidebar = "sidebar",
  Window = "window",
  ImageDetail = "imageDetail"
}

export interface IntentUISidebarIntegration
{
  anchor: IntentUiAnchor.Sidebar,
  isExternal: boolean
}

export interface IntentUIWindowIntegration
{
  anchor: IntentUiAnchor.Window;
}

export interface IntentUIModalIntegration
{
  anchor: IntentUiAnchor.Modal;
}

export type IntentUIIntegration =
  IntentUISidebarIntegration
  | IntentUIWindowIntegration
  | IntentUIModalIntegration;

export interface IntentUrlContent
{
  readonly url: string;
}

export interface IntentHtmlContent
{
  readonly html: string;
}

export type IntentFrameContent = IntentUrlContent | IntentHtmlContent;

export interface IntentUi
{
  readonly id: string;
  readonly integration: IntentUIIntegration;
  readonly frameContent: IntentFrameContent;
  readonly dialogContent?: IntentDialogIconContent;
}

export interface UiIntent extends WithContextIntent
{
  readonly ui: IntentUi;
}

export interface IntentOpenBrowser
{
  readonly url: string;
}

export interface OpenBrowserIntent extends BasisIntent
{
  readonly openBrowser: IntentOpenBrowser;
}

export enum IntentDialogType
{
  Error = "error",
  Info = "info",
  Question = "question"
}

export interface IntentFrame
{
  readonly content: IntentFrameContent;
  readonly height: number;
}

export interface IntentDialogButtons
{
  yes: string;
  no?: string;
}

export interface IntentDialog extends IntentDialogIconSizeContent
{
  readonly type: IntentDialogType;
  readonly frame?: IntentFrame;
  readonly buttons: IntentDialogButtons;
}

export interface DialogIntent extends WithContextIntent
{
  readonly dialog: IntentDialog;
}

export interface IntentImage
{
  readonly imageId: string;
  readonly dialogContent?: IntentDialogContent;
}

export interface IntentImages
{
  readonly images: IntentImage[];
  readonly dialogContent: IntentDialogIconContent;
}

export interface ImagesIntent extends WithContextIntent
{
  readonly images: IntentImages;
}

export enum IntentShowType
{
  Sidebar = "sidebar",
  ExtensionSettings = "extensionSettings",
  Image = "image",
  Repository = "repository"
}

export interface IntentShow
{
  readonly type: IntentShowType;
  readonly id: string;
}

export interface ShowIntent extends BasisIntent
{
  readonly show: IntentShow;
}

export enum IntentToastType
{
  Info = "info",
  Cancel = "cancel",
  Error = "error"
}

export interface IntentToast
{
  readonly type: IntentToastType;
  readonly title?: string;
  readonly subtitle: string;
}

export interface ToastIntent extends WithContextIntent
{
  readonly toast: IntentToast;
}

export interface IntentNotification
{
  readonly title: string;
  readonly subtitle: string;
  readonly body: string;
  readonly silent: boolean;
  readonly icon?: Buffer;
  readonly isNative: boolean;
}

export interface NotificationIntent extends WithContextIntent
{
  readonly notification: IntentNotification;
}

export interface IntentProcessCommand
{
  readonly extensionId: string;
  readonly commandId: string;
}

export interface ProcessCommandIntent extends WithContextIntent
{
  readonly processCommand: IntentProcessCommand;
}

export interface IntentAction
{
  readonly intent: ShowIntent | UiIntent | OpenBrowserIntent | ProcessCommandIntent;
  readonly dialogContent: IntentDialogIconSizeContent;
  readonly label?: string;
}

export interface ActionIntent extends WithContextIntent
{
  readonly action: IntentAction;
}

export type FrontIntent =
  FormIntent
  | UiIntent
  | OpenBrowserIntent
  | DialogIntent
  | ImagesIntent
  | ShowIntent
  | ToastIntent
  | NotificationIntent
  | ActionIntent;

export interface IntentServeBundle
{
  readonly content: Buffer;
  readonly settings?: Json;
}

export interface BundleIntent extends BasisIntent
{
  readonly serveBundle: IntentServeBundle;
}

export interface IntentReadFile
{
  readonly extensions?: string [];
  readonly message: string;
}

export interface ReadFileIntent extends WithContextIntent
{
  readonly readFile: IntentReadFile;
}

export interface IntentWriteFile
{
  readonly content: Buffer;
  readonly name: string;
  readonly extension: string;
  readonly message: string;
}

export interface WriteFileIntent extends WithContextIntent
{
  readonly writeFile: IntentWriteFile;
}

export type BackIntent =
  | BundleIntent
  | ReadFileIntent
  | WriteFileIntent;

export type Intent = FrontIntent | BackIntent;
